### PR TITLE
Split walkers among crowds in optimizer to match VMC usage

### PR DESCRIPTION
**Pull request by markdewing**
Addresses #3849 

The code now uses walker_per_crowd computed the same way as VMC.

Previously the batched cost function code expected every crowd would have the same number walkers. Most of the changes come from updating the code to use different number of walkers per crowd.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop, Summit

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:750]: If you prefer not to see the warning again. We can add an option to suppress warnings in adjustGlobalWalkerCount

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:273]: If this is a convenient variable, let us add a const.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:302]: If this is a convenient variable, let us add a const.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:517]: This may cause larger than expected imbalance when crowd sizes are not equal. Need to make sure per-walker number of samples differ by one at maximum instead of the number of per-crowd samples differ by 1.

File: src/QMCDrivers/tests/test_QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/tests/test_QMCCostFunctionBatched.cpp:144]: Does putting int directly work? I never find Index alias useful and  we probably should think about removing it.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:305]: This variable can't be const because it may get changed in the following lines.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:520]: I added a TODO comment to the same code in checkConfigurations.  Could this change be moved to a future PR?

Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:520]: Yes. But please open an issue to track this.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:285]: rename this vector. It is not samples_per_crowd but offsets.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:297]: Put this one as const.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:305]: I see. Then leave it not const. Could you rename this one as `this_batch_size`?

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:297]: This isn't part of the code that was changed in this PR!  It starts to feel excessively picky when minor changes get requested that aren't part of the changed code.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:285]: Will rename to `samples_per_crowd_offsets`

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:305]: How about `current_batch_size`?  I don't like using `this` in variable names because of the special meaning of "this" in C++.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: I should have said "could you please put this one as const?"
Existing code are subject to post-merge review as well.
When I read though the existing code trying to understand how `samples_per_crowd` is defined/used, I read this line and would like to have const added. That is also why I requested to rename the variable to reduce confusion.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: This is an example of surprise. Let me explain.
`curr_crowd_size` is somehow a const once its value is settled. So it is possible to made a decision and assign its value once and mark it const. It is better than set an initial value and then modify it later. Although the current way looks OK, the modification is on the second line but we do lost the const protection.

Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: `current_batch_size` is good

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:750]: You expressed concerns of printing duplicated warning. How would you like to handle this?
I'm actually fine with printing warning twice. First, the occurrence is low. If users decide to live with the warning, they can live with twice warning. Second, if we can reuse walker objects, then warning is also printed once only.
So I don't really feel this is critical. let me know your choice to add a warning suppression flag or leave it.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: Could you open an github issue? I think this the last thing blocking this PR (from my perspective).

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:750]: Printing the warning twice is fine for now.  If it becomes an issue later, we can add the flag.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: The ternary operator would be one solution.
```
const int current_batch_size = inb == num_batches - 1 ? final_batch_size : walkers_per_crowd[crowd_id
```

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: Opened #3888

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: extra consts added in several places.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: That is a clean solution that we can go for.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — COMMENTED** [review]
LGTM. Still need non-ANL approval.

**Review by prckent — APPROVED** [review]
Thanks Mark & Ye.

**Pull request merged**
